skimage/transform/tests/test_thin_plate_splines_transform.py

```diff
@@ -0,0 +1,164 @@
+import numpy as np
+import pytest
+from numpy.testing import assert_array_equal
+
+from skimage.transform._thin_plate_splines import TpsTransform, _ensure_2d, tps_warp
+
+SRC = np.array([[0, 0], [0, 5], [5, 5], [5, 0]])
+
+DST = np.array([[5, 0], [0, 0], [0, 5], [5, 5]])
+
+
+class TestTpsTransform:
+    def test_tps_transform_init(self):
+        tform = TpsTransform()
+
+        # Test that _estimated is initialized to False
+        assert tform._estimated is False
+        assert tform.coefficients is None
+        assert tform.src is None
+
+    def test_tps_transform_inverse(self):
+        tps = TpsTransform()
+        with pytest.raises(NotImplementedError):
+            tps.inverse()
+
+    def test_tps_transform_ensure_2d(self):
+        assert_array_equal(_ensure_2d(SRC), SRC)
+        assert_array_equal(_ensure_2d(DST), DST)
+
+        array_1d = np.array([0, 5, 10])
+        expected = np.array([[0], [5], [10]])
+        assert_array_equal(_ensure_2d(array_1d), expected)
+
+        empty_array = np.array([])
+        with pytest.raises(ValueError, match="Array of points can not be empty."):
+            _ensure_2d(empty_array)
+
+        scalar = 5
+        with pytest.raises(ValueError, match="Array must be be 2D."):
+            _ensure_2d(scalar)
+
+        array_3d = np.array([[[0, 5], [10, 15]], [[20, 25], [30, 35]]])
+        with pytest.raises(ValueError, match="Array must be be 2D."):
+            _ensure_2d(array_3d)
+
+        control_pts_less_than_3 = np.array([[0, 0], [0, 0]])
+        with pytest.raises(ValueError, match="Array points less than 3 is undefined."):
+            _ensure_2d(control_pts_less_than_3)
+
+    def test_tps_transform_estimation(self):
+        src = np.array([[0, 1], [-1, 0], [0, -1], [1, 0]])
+        dst = np.array([[0, 0.75], [-1, 0.25], [0, -1.25], [1, 0.25]], dtype=np.float32)
+        tform = TpsTransform()
+
+        # Ensure that the initial state is as expected
+        assert tform.coefficients is None
+        assert tform.src is None
+
+        # Perform estimation
+        tform.estimate(src, dst)
+        assert len(tform.src) > 0
+
+        assert len(tform.coefficients) > 0
+        assert tform.coefficients.shape[0] == src.shape[0] + 3
+
+        np.testing.assert_array_equal(tform.src, src)
+        np.testing.assert_allclose(
+            tform.coefficients[:, 0],
+            np.array([0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0]),
+            rtol=0.1,
+            atol=1e-16,
+        )
+        np.testing.assert_allclose(
+            tform.coefficients[:, 1],
+            np.array([-0.0902, 0.0902, -0.0902, 0.0902, 0.0, 0.0, 1.0]),
+            rtol=0.1,
+            atol=1e-16,
+        )
```

Comment:
Is there a reasoning behind splitting this into two separate checks instead of defining a single `desired_coefficients` array beforehand?

Reply:
Yes. According to Bookstein, 1989, page 571, the first test the coefficients for the x-coordinates and the second, those of the y-coordinates. However we can define a single array beforehand and test for x, y coordinates in one swoop